test/cpp/jit/test_profiling_pass.cpp

```diff
@@ -0,0 +1,270 @@
+      torch::jit::Operator(
+          FusedAddGraphSymbol,
+          createFusedAddGroup,
+          AliasAnalysisKind::PURE_FUNCTION),
+  });
+}
+
+struct ForEachAddFuser {
+  ForEachAddFuser(std::shared_ptr<Graph> graph) : graph_(std::move(graph)) {
+    ConstantPooling(graph_);
+    alias_db_ = std::make_unique<AliasDb>(graph_);
+  }
+
+ public:
+  Node* CreateFusedAddSubgraph(Node* n) {
+    GRAPH_UPDATE("Creating a fused ::Group node from: ", *n);
+    return SubgraphUtils::createSingletonSubgraphAndUpdateAliasing(
+        n, FusedAddGraphSymbol, *alias_db_);
+  }
+
+  bool checkTensorInputType(const TensorTypePtr& tt) {
+    // check input tensor conditions: here, just cuda
+    return tt->device()->is_cuda();
+  }
+
+  void createFusionGroups(Block* b) {
+    // iterate over Nodes in graph, trying horizontall merge add nodes with the
+    // same scalar
+    Node* curr_fusion_graph = nullptr;
+    Value* curr_scalar_arg = nullptr;
+    // NB: in this fusion pass, we are just merging nodes horizontally,
+    // typically, one might pull in inputs to fusion group to fuse vertically
+    // See tensorexpr_fuser.cpp or create_autodiff_subgraphs.cpp
+    // for example on vertical fusion
+
+    for (auto it = b->nodes().begin(); it != b->nodes().end();) {
+      Node* node = *it;
+      for (Block* block : node->blocks()) {
+        createFusionGroups(block);
+      }
+      it++; // advance iterator bc the current node may be destroyed
+      if (!node->matches(
+              "aten::add.Scalar(Tensor self, Scalar other, Scalar alpha=1) -> Tensor",
+              /*const_inputs=*/attr::alpha)) {
+        continue;
+      }
+      // _foreach_add.Scalar(Tensor[] tensors, Scalar scalar) -> Tensor[]
+      // does not support alpha
+      if (node->get<at::Scalar>(attr::alpha)->toDouble() != 1) {
+        continue;
+      }
+      // only supported on CUDA
+      if (!checkTensorInputType(node->input(0)->type()->expect<TensorType>())) {
+        continue;
+      }
+      Value* scalar_arg = node->input(1);
+      // if there exists a fusion graph, and we can merge current node after it,
+      // and the same scalar arg exists between the two, merge them
+      if (curr_fusion_graph && scalar_arg == curr_scalar_arg &&
+          alias_db_->moveAfterTopologicallyValid(node, curr_fusion_graph)) {
+        SubgraphUtils::mergeNodeIntoSubgraphAndUpdateAliasing(
+            node, curr_fusion_graph, *alias_db_);
+        continue;
+      }
+      curr_scalar_arg = scalar_arg;
+      curr_fusion_graph = CreateFusedAddSubgraph(node);
+      fusion_groups_.push_back(curr_fusion_graph);
+    }
+  }
+
+  void guardFusionGroups() {
+    for (Node* n : fusion_groups_) {
+      // the types present in the graph were present during profiling,
+      // and are not guaranteed in future executions
+      // to do this (cuda-only) optimization, we need to guard that the inputs
+      // are cuda If the guards fail, then we re-profile the fused graph nodes
```

Comment:
Is this true? I think in profiling_executor, `replaceFallbackGraphWithFallbackFunction()` isn't called after customPostPasses (afaik this is required in order for reprofiling to work)